### ISSUE TITLE
Inconsistent type inference

### ISSUE THREAD

**Issue by que-etc** [Microsoft/TypeScript#22472]
**TypeScript Version:** 2.8.0-dev.20180308

**Search Terms:**
* inferred type
* narrow

**Code**

```ts
const fn = (options: {foo: 'BAR'}) => {};

const fnOptions = {foo: 'BAR'};

// TS tells that "fnOptions" is not assignable to the "options" parameter.
fn(fnOptions);

// Although following invocation doesn't raise errors.
fn({foo: 'BAR'});
```

**Expected behavior:**
When passing `fnOptions` to `fn` function inferred type of its `foo` property should be narrowed to `"BAR"`.

**Actual behavior:**
At the time of `fn` invocation inferred type of `fnOptions` is `{foo: string}` which renders it not assignable to the `options` parameter.

**Playground Link:** http://www.typescriptlang.org/play/index.html#src=const%20fn%20%3D%20(options%3A%20%7Bfoo%3A%20'BAR'%7D)%20%3D%3E%20%7B%7D%3B%0D%0A%0D%0Aconst%20fnOptions%20%3D%20%7Bfoo%3A%20'BAR'%7D%3B%0D%0A%0D%0A%2F%2F%20TS%20tells%20that%20%22fnOptions%22%20is%20not%20assignable%20to%20the%20%22options%22%20parameter.%0D%0Afn(fnOptions)%3B%0D%0A%0D%0A%2F%2F%20Although%20following%20invocation%20doesn't%20raise%20errors.%0D%0Afn(%7Bfoo%3A%20'BAR'%7D)%3B

**Related Issues:** https://github.com/Microsoft/TypeScript/issues/17943
